finality-aleph/src/data_io.rs

```diff
@@ -140,214 +403,603 @@ where
+                    TopBlockImportedButNotFinalizedAncestor
+                }
+            }
+            TopBlockImportedButIncorrectBranch => {
+                // This could also be a hopeless fork, but we have checked before that it isn't (yet).
+                TopBlockImportedButIncorrectBranch
+            }
+            AvailableHopelessFork | AvailableBlock => {
+                error!(target: "afa", "Checking availability of an available proposal {:?} status: {:?}", proposal, old_status);
+                old_status.clone()
+            }
+        }
+    }
+
+    // For a proposal that might be new or not, check if it is available. If it is a new proposal
+    // and it is not available:
+    // 1) create an entry in pending_proposals
+    // 2) link this proposal to the input message by modifying the new added entry in pending proposals
+    //    and the provided message_info.
+    // 3) register an appropriate event trigger (for block import and/or finality).
+    // If the proposal is available, message_info is not modified.
+    fn add_message_proposal_dependency(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        message_info: &mut PendingMessageInfo<B, Message>,
+        id: MessageId,
+    ) {
+        if self.available_proposal_cache.contains(proposal) {
+            return;
+        }
+
+        if !self.pending_proposals.contains_key(proposal) {
+            use ProposalStatus::*;
+            let status = self.check_proposal_availability(proposal, &PendingTopBlock);
+            match status {
+                PendingTopBlock => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_block_import_trigger(proposal, &proposal.top_block());
+                    self.register_next_finality_trigger(proposal);
+                }
+                TopBlockImportedButIncorrectBranch => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_next_finality_trigger(proposal);
+                }
+                TopBlockImportedButNotFinalizedAncestor => {
+                    self.pending_proposals
+                        .insert(proposal.clone(), PendingProposalInfo::new(status));
+                    self.register_next_finality_trigger(proposal);
+                }
+                AvailableBlock | AvailableHopelessFork => {
+                    self.available_proposal_cache.put(proposal.clone(), ());
+                    // Proposal available, no need to register any dependencies
+                    return;
                 }
             }
         }
+        // This line is reached only if the proposal is not available
+        let proposal_info = self
+            .pending_proposals
+            .get_mut(proposal)
+            .expect("exists as checked above");
```

Comment:
You can use `Entry` to always return a mutref from the code above.

Reply:
Well, I just lost 10mins trying to do that, but encountered issues with the borrow checker. If you know how to do that, feel free to drop the code in a comment, otherwise I will leave it as it is.